server/controllers/conversation.js

```diff
@@ -1,122 +1,131 @@
-          await conversation.save();
-
-          res.status(200).json({
-              success: {
-                conversation: {
-                  name: existConversation.name,
-                  email: existConversation.email,
-              }
-            },
-          });
-        }
+    message = await Message.create({
+      senderId,
+      receiverId,
+      message: initialMessage,
     });
+  }
 
+  return res.status(201).send({
+    success: {
+      message,
+    },
+  });
+});
 
 // @route GET /conversation/getAllMessageByConversation
 // @desc get all messages from a single conversation
 // @access Public
- exports.getAllMessageByConversation = asyncHandler(async (req, res, next) => {
-
-    const conversations = await Conversation.find({
-      otherUsers: { $in: req.user.id },
-    }).populate({
-        path: "messages",
-        sort: { updatedAt: "desc" },
-      });
-
-    res.status(200).json({
-      success: {
-        message: conversations[0].messages,
-      },
-    });
-  
+exports.getAllMessageByConversation = asyncHandler(async (req, res, next) => {
+  const { receiverId } = req.params;
+
+  const conditions = {
+    senderId: mongoose.Types.ObjectId(req.user.id),
+    receiverId: mongoose.Types.ObjectId(receiverId),
+  };
+  const conversations = await getAllMessagesOnAConversation(conditions);
+
+  res.status(200).json({
+    success: {
+      messages: conversations,
+    },
+  });
 });
 
 // @route POST /conversation/sendMessage
 // @desc send a message to a conversation
 // @access Private
 exports.sendMessage = asyncHandler(async (req, res, next) => {
-    const { otherUsers, content  } = req.body;
-  
-    if (!content || !otherUsers) {
-      res.status(400).send("Bad request");
-    }
-
-    const conversation = await Conversation.findOne({
-      otherUsers:{ $all: [`${req.user.id}`, `${otherUsers}`]}
+  const { receiverId, message } = req.body;
+
+  const senderId = req.user.id;
+
+  const conversation = await Conversation.findOne({
+    receiverId: mongoose.Types.ObjectId(receiverId),
+    senderId: mongoose.Types.ObjectId(senderId),
+  });
+  let messages;
+
+  if (!conversation) {
+    await Conversation.create({
+      receiverId,
+      senderId,
     });
-    
-    const message = await Message.create({
-      sender: req.user.id,
-      content,
+
+    messages = await Message.create({
+      senderId,
+      receiverId,
+      message: message,
+    });
+  } else {
+    messages = await Message.create({
+      senderId,
+      receiverId,
+      message: message,
     });
+  }
 
-    conversation.messages.push(message);
+  res.status(200).send({
+    success: {
+      message: messages,
+    },
+  });
+});
 
-    await conversation.save();
+// @route GET /conversation/getAllConversations
+// @desc get all conversations for a user
+// @access Private
+exports.getAllConversations = asyncHandler(async (req, res, next) => {
+  const conditions = {
+    senderId: mongoose.Types.ObjectId(req.user.id),
```

Comment:
It's a good idea to make this conversion, but mongoose will always attempt to cast the provided `id` to object id type before querying the database.

Reply:
I tried earlier alone just passing in the id, but it could fail and the reason was just being seen as a normal string not an objectId

Reply:
Shouldn't have to do this, maybe look into why it's not working. I haven't ever done this ever in my experience with Mongoose.

Reply:
Could it possibly be that the senderId is referencing the Profile model but here we're passing a User object id. 🤔